python/lsst/ctrl/mpexec/job_report.py

```diff
@@ -0,0 +1,187 @@
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = ["ExecutionStatus", "JobReport", "QuantumReport"]
+
+import enum
+import shlex
+import sys
+from typing import Dict, List, Optional
+
+from lsst.daf.butler import DataCoordinate, DataId, DataIdValue
+from pydantic import BaseModel
+
+
+def _cmdLine() -> str:
+    """Return command line for current process."""
+    return shlex.join(sys.argv)
+
+
+def _serializeDataId(dataId: DataId) -> Dict[str, DataIdValue]:
+    if isinstance(dataId, DataCoordinate):
+        return dataId.byName()
+    else:
+        return dataId  # type: ignore
+
+
+class ExecutionStatus(enum.Enum):
+    """Possible values for job execution status.
+
+    Status `FAILURE` is set if one or more tasks failed. Status `TIMEOUT` is
+    set if there are no failures but one or more tasks timed out. Timeouts can
+    only be detected in multi-process mode, child task is killed on timeout
+    and usually should have non-zero exit code.
+    """
+
+    SUCCESS = "success"
+    FAILURE = "failure"
+    TIMEOUT = "timeout"
+
+
+class ExceptionInfo(BaseModel):
+    """Information about exception."""
+
+    className: str
+    """Name of the exception class if exception was raised."""
+
+    message: str
+    """Exception message for in-process quantum execution, None if
+    quantum was executed in sub-process.
+    """
+
+    def __init__(self, exception: Exception):
+        """Construct instance from an exception."""
+        super().__init__(className=exception.__class__.__name__, message=str(exception))
+
+
+class QuantumReport(BaseModel):
+    """Task execution report for a single Quantum."""
+
+    status: ExecutionStatus
+    """Execution status, one of the values in `ExecutionStatus` enum."""
+
+    dataId: Dict[str, DataIdValue]
+    """Quantum DataId."""
+
+    taskLabel: Optional[str]
+    """Label for a task executing this Quantum."""
+
+    exitCode: Optional[int] = None
+    """Exit code for a sub-process executing Quantum, None for in-process
+    Quantum execution. Negative if process was killed by a signal.
+    """
+
+    exceptionInfo: Optional[ExceptionInfo] = None
+    """Exception information if exception was raised."""
+
+    def __init__(
+        self,
+        dataId: DataId,
+        taskLabel: str,
+        status: ExecutionStatus = ExecutionStatus.SUCCESS,
+        exitCode: Optional[int] = None,
+        exception: Optional[Exception] = None,
+    ):
+        super().__init__(
+            status=status.value,
+            dataId=_serializeDataId(dataId),
```

Comment:
The usual way of doing this sort of thing in pydantic is to define validators: https://pydantic-docs.helpmanual.io/usage/validators/ and https://github.com/lsst/daf_butler/blob/main/python/lsst/daf/butler/core/serverModels.py#L200-L212

You use those to examine a specific parameter and modify it if needed.